Roll the dice example for elixir/phoenix

marcdel opened the pull request:
heavily cribbed from https://github.com/open-telemetry/opentelemetry.io/pull/2624

- [ ] squash these commits

---

Preview: https://deploy-preview-2685--opentelemetry.netlify.app/docs/instrumentation/erlang/getting-started/

chalin commented on content/en/docs/instrumentation/erlang/getting-started/rebar.md:
```diff
@@ -1,12 +1,9 @@
-title: Getting Started
-weight: 20
+title: Rebar3 / Mix
+aliases: [/docs/erlang/getting_started/rebar]
```
No need for an alias:

```suggestion
```

marcdel replied:
thanks! totally just cargo-culted these from the other doc

chalin commented on content/en/docs/instrumentation/erlang/getting-started/phoenix.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: Phoenix
+aliases: [/docs/erlang/getting_started/phoenix]
```
No need for an alias:

```suggestion
```

chalin commented on content/en/docs/instrumentation/erlang/getting-started/_index.md:
```diff
@@ -0,0 +1,18 @@
+---
+title: Getting Started
+aliases: [/docs/erlang/getting_started]
```
No need for an alias:

```suggestion
```

chalin reviewed (commented):
Very preliminary feedback: drop the `aliases` entries, they aren't needed.

tsloughter commented on content/en/docs/instrumentation/erlang/getting-started/phoenix.md:
````diff
@@ -0,0 +1,271 @@
+To begin, use the `phx` generator to create a new project
+
+```shell
+mix phx.new your_app_name
````
I think instead of `your_app_name` this should just use the example app name that'll go in the -contrib repo examples.

marcdel replied:
Question about this: there's a [basic_phoenix_ecto](https://github.com/open-telemetry/opentelemetry-erlang-contrib/tree/main/examples/basic_phoenix_ecto), I assume this would be a separate `dice` (`dice_game`?) example in addition to that one?

tsloughter replied:
Yup, a separate dice game example.

tsloughter commented on content/en/docs/instrumentation/erlang/getting-started/phoenix.md:
```diff
@@ -0,0 +1,271 @@
+  the API. Without it, all the functions in the API are no-ops.
+- `opentelemetry_exporter`: allows you to send your telemetry data to an
+  OpenTelemetry Collector and/or to self-hosted or commercial services.
+- `opentelemetry_ecto`: creates OpenTelemetry spans from the Elixir `:telemetry`
```
Ecto should be left out from the example since it isn't used. We want this to start without a database. But since it is likely any user will have Ecto it should definitely include a link and mention of `opentelemetry_ecto`.

Should also add a link to the demo, https://github.com/open-telemetry/opentelemetry-demo/tree/main/src/featureflagservice, and examples dir of contrib, https://github.com/open-telemetry/opentelemetry-erlang-contrib/tree/main/examples/, saying something like, 'for more complete Phoenix/Ecto projects see these'.

marcdel replied:
Should the example be explicitly created with `--no-ecto`?

tsloughter replied:
Yea.

tsloughter commented on content/en/docs/instrumentation/erlang/getting-started/phoenix.md:
```diff
@@ -0,0 +1,271 @@
+
+If you point your browser/curl/etc. to `http://localhost:4000/api/rolldice` you
+should get a random number in response, and a few spans in your console. I'll
+point out some of the most relevant bits.
```
Should give some details about how this is showing the span name and attributes for 2 spans.

tsloughter commented on content/en/docs/instrumentation/erlang/getting-started/phoenix.md:
````diff
@@ -0,0 +1,271 @@
+                             undefined}}
+```
+
+These are the raw structured logs that will get sent when you configure the
````
```suggestion
These are the raw Erlang records that will get serialized and sent when you configure the
```

tsloughter commented:
Had a few comments but overall this looks great.

oliverswitzer commented:
👏👏 thanks for doing this @marcdel!

tsloughter commented:
2The reason is in part from #587 where there needs to be a single attribute to expose the schema url for the instrumentation library. So need bbb


On Sun, May 14, 2023, at 14:37, Marc Delagrammatikas wrote:
> 
> 
> ***@***.**** commented on this pull request.
> 
> 
> In content/en/docs/instrumentation/erlang/getting-started/phoenix.md <https://github.com/open-telemetry/opentelemetry.io/pull/2685#discussion_r1193205542>:
> 
> > +
> +ℹ️ Throughout this guide I will use `your_app_name` or `YourAppName`, be sure to
> +replace these with your actual application's name.
> +
> +### Dependencies
> +
> +We'll need a few other dependencies that Phoenix doesn't come with as well.
> +
> +- `opentelemetry_api`: contains the interfaces you'll use to instrument your
> +  code. Things like `Tracer.with_span` and `Tracer.set_attribute` are defined
> +  here.
> +- `opentelemetry`: contains the SDK that implements the interfaces defined in
> +  the API. Without it, all the functions in the API are no-ops.
> +- `opentelemetry_exporter`: allows you to send your telemetry data to an
> +  OpenTelemetry Collector and/or to self-hosted or commercial services.
> +- `opentelemetry_ecto`: creates OpenTelemetry spans from the Elixir `:telemetry`
> 
> Should the example be explicitly created with `--no-ecto`?
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/open-telemetry/opentelemetry.io/pull/2685#discussion_r1193205542>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAI3A3P7BYIP3C56CZHKHLXGE673ANCNFSM6AAAAAAX4J7WRM>.
> You are receiving this because your review was requested.Message ID: ***@***.***>
>

svrnm commented:
@marcdel, @tsloughter not knowing enough about erlang/elixir, so I leave that to you. I have one question/concern I'd like to clarify: One of the goals we try to accomplish with standardizing on one application for "Getting Started" is to get consistency across the docs. While this PR accomplishes that partially I still see some inconsistencies I'd like to see if we can resolve them:

![Screenshot 2023-05-15 at 10 41 16](https://github.com/open-telemetry/opentelemetry.io/assets/1519757/b2c234e7-bfc7-4273-b360-23f3ec0e9553)

There are now 2 kinds of getting started, which is confusing. Would it be possible to make the "Phoenix" one with the roll the dice app the REAL getting started page and take whatever is important from the existing Getting Started and merge it and put everything else into pages where it fits (e.g. exporters have their own page across the languages)

Also, if I look at the Phoenix app there is some differences to other languages. Would it be possible to structure the page in a similar way (e.g. start with a ready to use roll the dice app and instrument from there)

**Note**: We can accomplish some/all of that in follow-up PRs but I wanted to clarify that.

 cc @cartermp @chalin @austinlparker

tsloughter commented:
@svrnm ok, yea, I agree about starting from a working roll dice app and then instrument that. The README.md of the example can have any setup instructions about how the project was created. What do you think @marcdel ?

I think we can also move stuff into a single `Getting Started` and add some stuff to the index page about adding deps. The main use is certainly going to be adding to existing projects, and there are plenty of other guides on creating Erlang and Elixir projects.

svrnm commented:
@marcdel this is really great work so I am eager to merge this in, please let us know if and where you might need assistance, if any of our comments or change requests are unclear/a too big ask, we can clarify and help with the heavy lifting.

marcdel commented:
> **Note**: We can accomplish some/all of that in follow-up PRs but I wanted to clarify that.

> There are now 2 kinds of getting started, which is confusing. Would it be possible to make the "Phoenix" one with the roll the dice app the REAL getting started page and take whatever is important from the existing Getting Started and merge it and put everything else into pages where it fits (e.g. exporters have their own page across the languages)

@svrnm Does it make sense to combine these two pages and get this merged in, then address those other bits? I'd rather add pages than remove ones someone might already be linking to.

FWIW I was following the [javascript](https://deploy-preview-2685--opentelemetry.netlify.app/docs/instrumentation/js/getting-started/) one which has node as a child of the getting started page. I don't have any strong feelings on organization so I'm happy to do whatever!

marcdel commented:
Okie dokie, I've moved the Phoenix stuff into the main Getting Started page and shuffled some things around such that it hopefully makes sense. I also moved the setup bits over to the readme of https://github.com/open-telemetry/opentelemetry-erlang-contrib/pull/173, and edited the surrounding bits to assume all the setup is already done.

tsloughter commented:
Thanks. I think this will work for now and I'll merge together the Phoenix and Erlang getting started once I write the Erlang part. So it won't have the section about creating a new project anymore once I do that.

cartermp reviewed (approved):
Thanks @marcdel, this is great. I'll merge this assuming nobody else has any objections in a day or so.
